typed/remove_test.go

```diff
@@ -0,0 +1,133 @@
+package typed
+
+import (
+	"fmt"
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v3/value"
+)
+
+func TestRemove(t *testing.T) {
+
+	var cases = []struct {
+		object YAMLObject
+		remove *fieldpath.Set
+		deep   bool
+		expect YAMLObject
+	}{
+		{
+			object: `{}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: ``,
+		},
+		{
+			object: `{"a": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: `null`,
+		},
+		{
+			object: `{"a": "value", "b": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: `{"b": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: `{"b": {}}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			deep:   true,
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b":{"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			deep:   false,
+			// TODO: this seems to behave differently with a non deduced schema: {"a": "value", "b":{"c":"value"}} which is the reason for this PR
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b", "c")),
+			deep:   true,
+			// TODO: is a null result expected (insted of a {})
```

Comment:
I agree that it's a little odd. What do you think @jennybuckley

Reply:
same reason as: https://github.com/kubernetes-sigs/structured-merge-diff/pull/133/files#r364479060

Reply:
I think this wouldn't matter because in an actual serialized kubernetes object with an empty map, it should look the same as one with a null map, right? but I'm not totally sure

Reply:
Not sure. Don't we treat null and empty differently at least for apply?

Reply:
So when we strip fields before running apply, I think resulting in null actually might be better as the key would be removed from the object and apply ignores it (right?).
Probably the reason why it is this way?

Reply:
Actually, we tend to treat null as an opinion so it's not ideal

Reply:
Do we? The opinion to remove something/reset a list?
I only remember when I added the tests for resetting managedFields, that the managedFields must not be null or [] but [{}] but I think that's a different place.

Reply:
You're right, we don't on list, we should be good then.